core/src/org/samtools/htsjdk/core/io/ByteSource.java

```diff
@@ -0,0 +1,107 @@
+package org.samtools.htsjdk.core.io;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Interface for classes that produce binary data (i.e. bytes) that can be decoded
+ * into other datatypes.
+ *
+ * Methods should throw exceptions if:
+ * - The underlying data source/stream/channel is closed
+ * - There are insufficient bytes available (after blocking) to read a single item of the datatype requested
+ *
+ * NOTES:
+ *   - Almost all the methods here could have default implementations that read via one of the readBytes() methods.
+ *   - Should think through how to deal with unsigned types
+ *   - Need to think how this would work with BlockCompressed data
+ */
+public interface ByteSource extends AutoCloseable {
+    /** Returns a description of the underlying source of the bytes, e.g. a path, stdin, a URL etc. */
+    String getDescription();
+
+    // Methods related to buffering/repositioning/skipping etc.
+    /** Returns true if the source has a known finite length, false otherwise. */
+    boolean hasLength();
+
+    /** Returns the length if {@link #hasLength()} is true, otherwise throws an exception. */
+    default long length() { throw new UnsupportedOperationException(); }
+
+    /** Returns the offset of the next byte that will be read by any operation. */
+    long position();
+
+    /** Returns true if the source can skip bytes without reading them. */
+    boolean canNavigateForward();
+
+    /** Returns true if the source can always be repositioned behind the current position. Note that a buffered
+     * source may allow limited backwards navigation within the buffered data, while returning false here. */
+    boolean canNavigateBackwards();
```

Comment:
I might call this `canSeek` and `navigateToPosition`-> `seek(long)`.  Is there any input that supports reverse reading but not random access?  Symmetry is nice, but maybe in this case it's not necessary?

Reply:
I'm not sure what the behavior should be for buffered readers that have some amount of look behind.  We could use mark/reset but that's always kind of awkward.  We could have some method like `Range getSeekableRange()` that returns the range you're allowed to seek within.

Reply:
I agree, I find `mark()` and `reset()` awkward.  Perhaps just a `canSeekTo(long position)` would be easiest?  Because I think, as a client, that's really what I want to know.  I.e. can `canSeekTo(0)`?